graphs/gen/holme_kim.go

```diff
@@ -0,0 +1,156 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gen
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+
+	"github.com/gonum/graph"
+	"github.com/gonum/graph/simple"
+	"github.com/gonum/stat/sample"
+)
+
+// TunableClusteringScaleFree constructs a graph in the destination, dst, of order n.
+// At each iteration of graph addition, one node is added with m additional edges
+// joining existing nodes with probability proportional to the nodes' degrees. The
+// edges are formed as a triad with probability, p. If src is not nil it is used as
+// the random source, otherwise rand.Float64 and rand.Intn are used.
+//
+// The algorithm is essentially as described in http://arxiv.org/abs/cond-mat/0110452.
+func TunableClusteringScaleFree(dst graph.UndirectedBuilder, n, m int, p float64, src *rand.Rand) error {
+	if p < 0 || p > 1 {
+		return fmt.Errorf("gen: bad probability: p=%v", p)
+	}
+	if n < m {
```

Comment:
Good catch. It should be here; we can't make an order n graph with each node degree n in a simple graph.